File: core/src/com/spacefist/weapons/LaserWeapon.java
package com.spacefist.weapons;

import com.spacefist.GameData;
import com.spacefist.entities.Ship;
import com.spacefist.managers.ProjectileManager;
import com.spacefist.weapons.abst.Weapon;
import org.jetbrains.annotations.NotNull;

/**
 * The default weapon
 */
public class LaserWeapon implements Weapon {
    private ProjectileManager projectileManager;

    private GameData gameData;

    /**
      * Creates a new LaserWeapon instance
      *
      * @param gameData Common game data
      */
    @SuppressWarnings("UnnecessaryThis")
    public LaserWeapon(@NotNull GameData gameData) {
        this.gameData = gameData;

        this.projectileManager = gameData.getProjectileManager();
    }

    @Override
    public void fire() {
        Ship ship = gameData.getShip();

        float shipWidth = ship.getRectangle().getWidth();
        float shipHeight = ship.getRectangle().getHeight();

        int projectileX = (int) (ship.getX() + (shipWidth / 2));
        int projectileY = (int) ((ship.getY() + shipHeight));

        projectileManager.fireLaser(projectileX, projectileY);
    }
}

File: core/src/com/spacefist/weapons/LaserWeapon.kt
package com.spacefist.weapons

import com.spacefist.GameData
import com.spacefist.entities.Ship
import com.spacefist.managers.ProjectileManager
import com.spacefist.weapons.abst.Weapon

/**
 * The default weapon
 */
class LaserWeapon
/**
 * Creates a new LaserWeapon instance
 *
 * @param gameData Common game data
 */
(private val gameData: GameData) : Weapon {
    private val projectileManager: ProjectileManager

    init {

        this.projectileManager = gameData.projectileManager
    }

    override fun fire() {
        val ship = gameData.ship

        val shipWidth = ship.rectangle.getWidth()
        val shipHeight = ship.rectangle.getHeight()

        val projectileX = (ship.x + shipWidth / 2).toInt()
        val projectileY = (ship.y + shipHeight).toInt()

        projectileManager.fireLaser(projectileX, projectileY)
    }
}
